hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/clean/CleanPlanner.java

```diff
@@ -472,4 +488,43 @@ private boolean isFileSliceNeededForPendingCompaction(FileSlice fileSlice) {
   private boolean isFileGroupInPendingCompaction(HoodieFileGroup fg) {
     return fgIdToPendingCompactionOperations.containsKey(fg.getFileGroupId());
   }
+
+  public boolean isDeletePartitionOperation(Option<HoodieInstant> instantToRetain) {
+    try {
+      if (instantToRetain.isPresent() && HoodieTimeline.REPLACE_COMMIT_ACTION.equals(instantToRetain.get().getAction())) {
+        HoodieReplaceCommitMetadata replaceCommitMetadata = HoodieReplaceCommitMetadata.fromBytes(
+            hoodieTable.getActiveTimeline().getInstantDetails(instantToRetain.get()).get(), HoodieReplaceCommitMetadata.class);
+
+        if (replaceCommitMetadata != null
+            && WriteOperationType.DELETE_PARTITION.equals(replaceCommitMetadata.getOperationType())) {
+          return true;
+        }
+      }
+    } catch (Exception e) {
+      throw new HoodieException("Failed to get commit metadata", e);
+    }
+    return false;
+  }
+
+  public List<String> getDropPartitionPathsForClean(Option<HoodieInstant> earliestRetainedInstant) {
+    List<String> dropPartitions = new ArrayList<>();
+    if (!earliestRetainedInstant.isPresent()) {
+      LOG.info("No earliest commit to retain. No need to scan partitions !!");
+      return dropPartitions;
+    }
+
+    try {
+      if (isDeletePartitionOperation(earliestRetainedInstant)) {
```

Comment:
guess we can't just check the earliestRetainedInstant alone here. 
for the same reason quoted above. we have to check for all instants prior to this from last cleaned up instant. 
can we revisit this logic please.